File: src/main/java/com/linux/presentation/JobsView.java
package com.linux.presentation;

import com.linux.auth.control.AuthenticationControl;
import com.linux.contact.control.ContactDataControl;
import com.linux.jobs.control.JobDataControl;
import com.linux.jobs.entity.Job;
import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.grid.Grid;
import com.vaadin.flow.component.orderedlayout.*;
import com.vaadin.flow.component.textfield.TextField;
import com.vaadin.flow.data.value.ValueChangeMode;
import com.vaadin.flow.router.*;
import io.quarkus.narayana.jta.QuarkusTransaction;
import org.jboss.logging.Logger;

import javax.inject.Inject;
import java.util.*;
import java.util.stream.*;

import static com.linux.presentation.JobForm.*;

@Route(value = "", layout = MainView.class)
@PageTitle(value = "Jobs")
public class JobsView extends VerticalLayout {
    private static final Logger l = Logger.getLogger(JobsView.class);

    private final Grid<Job> grid = new Grid<>();
    private final TextField filterText = new TextField();
    private final JobDataControl jobDataControl;
    private final ContactDataControl contactDataControl;
    private final AuthenticationControl authenticationControl;

    private JobForm jobForm;

    @Inject
    public JobsView(JobDataControl jobDataControl, ContactDataControl contactDataControl, AuthenticationControl authenticationControl) {
        this.jobDataControl = jobDataControl;
        this.contactDataControl = contactDataControl;
        this.authenticationControl = authenticationControl;
        addClassName("jobs-list");
        setSizeFull();
        updateList();
        configureGrid();
        configureForm();
        add(getToolbar(), getContent());
        closeEditor();
    }

    private void configureGrid() {
        grid.addClassName("jobs-grid");
        grid.setSizeFull();
        grid.addColumn((job) -> job.company).setHeader("Company").setSortable(true).setFrozen(true);
        grid.addColumn((job) -> job.currentJobOpenings).setHeader("Information About The Opening").setSortable(true).setFrozen(true);
        grid.addColumn((job) -> job.domains.stream().map((jobDomain) -> jobDomain.domain).collect(Collectors.joining(","))).setHeader("Job Domains").setSortable(true).setFrozen(true);
        grid.addColumn((job) -> job.slots.dbData).setHeader("Number Of Openings").setSortable(true).setFrozen(true);
        grid.addColumn((job) -> job.contact.name()).setHeader("Contact Information").setSortable(true).setFrozen(true);
        grid.getColumns().forEach(col -> col.setAutoWidth(true));
        grid.asSingleSelect().addValueChangeListener(event -> editJob(event.getValue()));
    }

    private HorizontalLayout getToolbar() {
        filterText.setPlaceholder("Filter by company...");
        filterText.setClearButtonVisible(true);
        filterText.setValueChangeMode(ValueChangeMode.LAZY);
        filterText.addValueChangeListener(e -> updateList());
        filterText.setValue("");

        Button addJobButton = new Button("Add Job");
        addJobButton.addClickListener(click -> addJob());

        HorizontalLayout toolbar = new HorizontalLayout(filterText, addJobButton);
        toolbar.addClassName("toolbar");
        return toolbar;
    }

    private void updateList() {
        filterText.getValue();
        grid.setItems(jobDataControl.findByCompany(filterText.getValue()));
    }

    private Component getContent() {
        HorizontalLayout content = new HorizontalLayout(grid, jobForm);
        content.setFlexGrow(2, grid);
        content.setFlexGrow(1, jobForm);
        content.addClassNames("content");
        content.setSizeFull();
        return content;
    }

    private void configureForm() {
        jobForm = new JobForm();
        jobForm.setWidth("25em");
        jobForm.addListener(SaveEvent.class, this::saveJob);
        jobForm.addListener(DeleteEvent.class, this::deleteJob);
        jobForm.addListener(CloseEvent.class, e -> closeEditor());
    }

    public void editJob(Job job) {
        if (job == null) {
            closeEditor();
            return;
        }
        if (job.contact == null || Objects.equals(contactDataControl.findByEmail(authenticationControl.email()).id, job.contact.id)) {
            jobForm.setJob(job);
            jobForm.setVisible(true);
            addClassName("editing");
        }
    }

    private void closeEditor() {
        jobForm.setJob(null);
        jobForm.setVisible(false);
        removeClassName("editing");
    }

    private void addJob() {
        grid.asSingleSelect().clear();
        editJob(new Job());
    }

    private void saveJob(SaveEvent event) {
        QuarkusTransaction.run(() -> {
            var job = event.job;
            if(!jobDataControl.isPersistent(job)) {
                l.warn("job from event is not persistent!" );
            }
            job.contact = contactDataControl.findOrCreate(authenticationControl.email());
            jobDataControl.persist(job);
            updateList();
            closeEditor();
        });
    }

    private void deleteJob(DeleteEvent event) {
        QuarkusTransaction.run(() -> jobDataControl.deleteById(event.job.id));
        updateList();
        closeEditor();
    }
}

File: src/main/kotlin/com/linux/presentation/JobsView.kt
package com.linux.presentation

import com.linux.auth.control.AuthenticationControl
import com.linux.contact.control.ContactDataControl
import com.linux.jobs.control.JobDataControl
import com.linux.jobs.entity.Job
import com.vaadin.flow.component.button.Button
import com.vaadin.flow.component.grid.Grid
import com.vaadin.flow.component.orderedlayout.HorizontalLayout
import com.vaadin.flow.component.orderedlayout.VerticalLayout
import com.vaadin.flow.component.textfield.TextField
import com.vaadin.flow.data.value.ValueChangeMode
import com.vaadin.flow.router.PageTitle
import com.vaadin.flow.router.Route
import io.quarkus.narayana.jta.QuarkusTransaction
import io.quarkus.oidc.IdToken
import org.eclipse.microprofile.jwt.JsonWebToken
import org.jboss.logging.Logger
import javax.annotation.PostConstruct
import javax.inject.Inject

@Suppress("unused")
@Route(value = "", layout = MainView::class)
@PageTitle("Jobs")
class JobsView(
    @Inject @IdToken var idToken: JsonWebToken,
    @Inject var authControl: AuthenticationControl,
    @Inject var jobDataControl: JobDataControl,
    @Inject var contactDataControl: ContactDataControl
) : VerticalLayout() {
    private val l = Logger.getLogger(JobsView::class.java)
    private final val grid = createGrid()
    private final val jobForm = createJobForm()

    private final val companyNameFilter = createCompanyNameFilter()

    @PostConstruct
    fun postConstruct() {
        setSizeFull()
        add(HorizontalLayout(companyNameFilter, Button("Add Job") { grid.asSingleSelect().clear(); onRowSelected(Job())}).apply { addClassName("toolbar") },
            HorizontalLayout(grid, jobForm).apply {
                addClassName("content")
                setFlexGrow(2.0, grid)
                setFlexGrow(1.0, jobForm)
                setSizeFull()
            })
        addClassName("jobs-list")

        grid.asSingleSelect().addValueChangeListener { event -> onRowSelected(event.value) }
    }

    private fun checkOrCreateContact() {
//        if (contactDataControl.contactAbsent(idToken.name)) {
//            QuarkusTransaction.run { contactDataControl.persist(Contact(email = idToken.name, givenName = idToken.getClaim("givenName"), familyName = idToken.getClaim("familyName"))) }
//        }
    }


    private fun closeEditor() {
        jobForm.jobSupplier = { null }
        jobForm.isVisible = false
        jobForm.removeClassName("editing")
    }


    private fun updateList() {
        if (companyNameFilter.value.isBlank()) {
            grid.setItems(jobDataControl.listAll())
        } else {
            grid.setItems(jobDataControl.list("company", companyNameFilter.value))
        }
    }

    private fun saveEvent(job: Job) {
        l.info("Save Event Called $job")
        QuarkusTransaction.run {
            job.contact = contactDataControl.findByEmail(authControl.email())!!
            l.info("Contact Id : ${job.contact.id}")
            jobDataControl.persist(job)
        }
        updateList()
        closeEditor()
    }

    private fun deleteEvent(job: Job) {
        l.info("Delete Event Called $job")
        job.contact = contactDataControl.findByEmail(authControl.email())!!
        jobDataControl.delete(job)
    }


    private fun createJobForm() = JobForm(this::saveEvent, this::deleteEvent, this::closeEditor).apply {
        width = "25em"
    }


    private fun createGrid() = Grid<Job>().apply {
        addClassName("job-grid")
        setSizeFull()
        addColumn { it.company }.setHeader("Company").apply { isSortable = true; isFrozen = true }
        addColumn { it.currentJobOpenings }.setHeader("Information About The Opening").apply { isSortable = true; isFrozen = true }
        addColumn { it.domains.joinToString { "," } }.setHeader("Job Domains").apply { isSortable = true; isFrozen = true }
        addColumn { it.slots.value }.setHeader("Number Of Openings").apply { isSortable = true; isFrozen = true }
        addColumn { it.contact.name() }.setHeader("Contact Information").apply { isFrozen = true }
        columns.forEach { it.setAutoWidth(true) }
        asSingleSelect().addValueChangeListener { it.value?.apply { onRowSelected(it.value) } }
    }

    private fun createCompanyNameFilter() = TextField().apply {
        placeholder = "Filter by Company"
        isClearButtonVisible = true
        valueChangeMode = ValueChangeMode.LAZY
    }

    private fun onRowSelected(job: Job) {
        jobForm.addClassName("editing")
        jobForm.isVisible = true
        jobForm.jobSupplier = { job }
    }
}